// payload/dist/express/middleware/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _bodyparser = /*#__PURE__*/ _interop_require_default(require("body-parser"));
const _compression = /*#__PURE__*/ _interop_require_default(require("compression"));
const _express = /*#__PURE__*/ _interop_require_default(require("express"));
const _expressfileupload = /*#__PURE__*/ _interop_require_default(require("express-fileupload"));
const _expressratelimit = /*#__PURE__*/ _interop_require_default(require("express-rate-limit"));
const _methodoverride = /*#__PURE__*/ _interop_require_default(require("method-override"));
const _passport = /*#__PURE__*/ _interop_require_default(require("passport"));
const _qsmiddleware = /*#__PURE__*/ _interop_require_default(require("qs-middleware"));
const _middleware = /*#__PURE__*/ _interop_require_default(require("../../localization/middleware"));
const _authenticate = /*#__PURE__*/ _interop_require_default(require("./authenticate"));
const _convertPayload = /*#__PURE__*/ _interop_require_default(require("./convertPayload"));
const _corsHeaders = /*#__PURE__*/ _interop_require_default(require("./corsHeaders"));
const _defaultPayload = /*#__PURE__*/ _interop_require_default(require("./defaultPayload"));
const _i18n = require("./i18n");
const _identifyAPI = /*#__PURE__*/ _interop_require_default(require("./identifyAPI"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const middleware = (payload)=>{
    const rateLimitOptions = {
        max: payload.config.rateLimit.max,
        windowMs: payload.config.rateLimit.window
    };
    if (typeof payload.config.rateLimit.skip === 'function') rateLimitOptions.skip = payload.config.rateLimit.skip;
    if (payload.config.express.middleware?.length) {
        payload.logger.warn('express.middleware is deprecated. Please migrate to express.postMiddleware.');
    }
    return [
        _defaultPayload.default,
        ...payload.config.express.preMiddleware || [],
        (0, _expressratelimit.default)(rateLimitOptions),
        _passport.default.initialize(),
        (0, _i18n.i18nMiddleware)(payload.config.i18n),
        (0, _identifyAPI.default)('REST'),
        (0, _methodoverride.default)('X-HTTP-Method-Override'),
        (0, _qsmiddleware.default)({
            arrayLimit: 1000,
            depth: 10,
            strictNullHandling: true
        }),
        _bodyparser.default.urlencoded({
            extended: true
        }),
        (0, _compression.default)(payload.config.express.compression),
        _middleware.default,
        _express.default.json(payload.config.express.json),
        (0, _expressfileupload.default)({
            parseNested: true,
            ...payload.config.upload
        }),
        _convertPayload.default,
        (0, _authenticate.default)(payload.config),
        (0, _corsHeaders.default)(payload.config),
        ...payload.config.express.middleware || [],
        ...payload.config.express.postMiddleware || []
    ];
};
const _default = middleware;
