// payload/dist/admin/components/views/collections/Edit/Default/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "DefaultCollectionEdit", {
    enumerable: true,
    get: function() {
        return DefaultCollectionEdit;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../../../utilities/getTranslation");
const _DocumentControls = require("../../../../elements/DocumentControls");
const _DocumentFields = require("../../../../elements/DocumentFields");
const _LeaveWithoutSaving = require("../../../../modals/LeaveWithoutSaving");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../../../utilities/Meta"));
const _Auth = /*#__PURE__*/ _interop_require_default(require("../Auth"));
const _SetStepNav = require("../SetStepNav");
const _Upload = require("../Upload");
const _formatFields = /*#__PURE__*/ _interop_require_default(require("../formatFields"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'collection-default-edit';
const DefaultCollectionEdit = (props)=>{
    const { i18n, t } = (0, _reacti18next.useTranslation)('general');
    const { id, apiURL, collection, data, disableActions, disableLeaveWithoutSaving, fieldTypes, hasSavePermission, internalState, isEditing, onSave, permissions } = props;
    const { auth, upload } = collection;
    const [fields] = (0, _react.useState)(()=>(0, _formatFields.default)(collection, isEditing));
    const operation = isEditing ? 'update' : 'create';
    return /*#__PURE__*/ _react.default.createElement(_react.Fragment, null, /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: `${isEditing ? t('editing') : t('creating')} - ${(0, _getTranslation.getTranslation)(collection.labels.singular, i18n)}`,
        keywords: `${(0, _getTranslation.getTranslation)(collection.labels.singular, i18n)}, Payload, CMS`,
        title: `${isEditing ? t('editing') : t('creating')} - ${(0, _getTranslation.getTranslation)(collection.labels.singular, i18n)}`
    }), !(collection.versions?.drafts && collection.versions?.drafts?.autosave) && !disableLeaveWithoutSaving && /*#__PURE__*/ _react.default.createElement(_LeaveWithoutSaving.LeaveWithoutSaving, null), /*#__PURE__*/ _react.default.createElement(_SetStepNav.SetStepNav, {
        collection: collection,
        id: id,
        isEditing: isEditing
    }), /*#__PURE__*/ _react.default.createElement(_DocumentControls.DocumentControls, {
        apiURL: apiURL,
        collection: collection,
        data: data,
        disableActions: disableActions,
        hasSavePermission: hasSavePermission,
        id: id,
        isEditing: isEditing,
        onSave: onSave,
        permissions: permissions
    }), /*#__PURE__*/ _react.default.createElement(_DocumentFields.DocumentFields, {
        BeforeFields: /*#__PURE__*/ _react.default.createElement(_react.Fragment, null, auth && /*#__PURE__*/ _react.default.createElement(_Auth.default, {
            className: `${baseClass}__auth`,
            collection: collection,
            email: data?.email,
            operation: operation,
            readOnly: !hasSavePermission,
            requirePassword: !isEditing,
            useAPIKey: auth.useAPIKey,
            verify: auth.verify
        }), upload && /*#__PURE__*/ _react.default.createElement(_Upload.Upload, {
            collection: collection,
            internalState: internalState
        })),
        fieldTypes: fieldTypes,
        fields: fields,
        hasSavePermission: hasSavePermission,
        permissions: permissions
    }));
};
