// @payloadcms/richtext-slate/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ElementButton: function() {
        return _Button.default;
    },
    LeafButton: function() {
        return _Button1.default;
    },
    nodeIsTextNode: function() {
        return _types.nodeIsTextNode;
    },
    slateEditor: function() {
        return slateEditor;
    },
    toggleElement: function() {
        return _toggle.default;
    }
});
const _utilities = require("payload/utilities");
const _cell = /*#__PURE__*/ _interop_require_default(require("./cell"));
const _richTextRelationshipPromise = require("./data/richTextRelationshipPromise");
const _validation = require("./data/validation");
const _field = /*#__PURE__*/ _interop_require_default(require("./field"));
const _Button = /*#__PURE__*/ _interop_require_default(require("./field/elements/Button"));
const _toggle = /*#__PURE__*/ _interop_require_default(require("./field/elements/toggle"));
const _Button1 = /*#__PURE__*/ _interop_require_default(require("./field/leaves/Button"));
const _types = require("./types");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function slateEditor(args) {
    return {
        CellComponent: (0, _utilities.withMergedProps)({
            Component: _cell.default,
            toMergeIntoProps: args
        }),
        FieldComponent: (0, _utilities.withMergedProps)({
            Component: _field.default,
            toMergeIntoProps: args
        }),
        outputSchema: ({ isRequired })=>{
            return {
                type: (0, _utilities.withNullableJSONSchemaType)('array', isRequired),
                items: {
                    type: 'object'
                }
            };
        },
        populationPromise ({ context, currentDepth, depth, draft, field, findMany, flattenLocales, overrideAccess, populationPromises, req, showHiddenFields, siblingDoc }) {
            if (field.admin?.elements?.includes('relationship') || field.admin?.elements?.includes('upload') || field.admin?.elements?.includes('link') || !field?.admin?.elements) {
                return (0, _richTextRelationshipPromise.richTextRelationshipPromise)({
                    context,
                    currentDepth,
                    depth,
                    draft,
                    field,
                    findMany,
                    flattenLocales,
                    overrideAccess,
                    populationPromises,
                    req,
                    showHiddenFields,
                    siblingDoc
                });
            }
            return null;
        },
        validate: _validation.richTextValidate
    };
}
