// @payloadcms/richtext-slate/dist/field/elements/link/utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    transformExtraFields: function() {
        return transformExtraFields;
    },
    unwrapLink: function() {
        return unwrapLink;
    },
    withLinks: function() {
        return withLinks;
    },
    wrapLink: function() {
        return wrapLink;
    }
});
const _slate = require("slate");
const _baseFields = require("./LinkDrawer/baseFields");
const unwrapLink = (editor)=>{
    _slate.Transforms.unwrapNodes(editor, {
        match: (n)=>_slate.Element.isElement(n) && n.type === 'link'
    });
};
const wrapLink = (editor)=>{
    const { selection } = editor;
    const isCollapsed = selection && _slate.Range.isCollapsed(selection);
    const link = {
        children: isCollapsed ? [
            {
                text: ''
            }
        ] : [],
        newTab: false,
        type: 'link',
        url: undefined
    };
    if (isCollapsed) {
        _slate.Transforms.insertNodes(editor, link);
    } else {
        _slate.Transforms.wrapNodes(editor, link, {
            split: true
        });
        _slate.Transforms.collapse(editor, {
            edge: 'end'
        });
    }
};
const withLinks = (incomingEditor)=>{
    const editor = incomingEditor;
    const { isInline } = editor;
    editor.isInline = (element)=>{
        if (element.type === 'link') {
            return true;
        }
        return isInline(element);
    };
    return editor;
};
function transformExtraFields(customFieldSchema, config, i18n) {
    const baseFields = (0, _baseFields.getBaseFields)(config);
    const fields = typeof customFieldSchema === 'function' ? customFieldSchema({
        config,
        defaultFields: baseFields,
        i18n
    }) : baseFields;
    // Wrap fields which are not part of the base schema in a group named 'fields' - otherwise they will be rendered but not saved
    const extraFields = [];
    for(let i = fields.length - 1; i >= 0; i--){
        const field = fields[i];
        if ('name' in field) {
            if (!baseFields.find((baseField)=>!('name' in baseField) || baseField.name === field.name)) {
                if (field.name !== 'fields' && field.type !== 'group') {
                    extraFields.push(field);
                    // Remove from fields from now, as they need to be part of the fields group below
                    fields.splice(fields.indexOf(field), 1);
                }
            }
        }
    }
    if (Array.isArray(customFieldSchema) || fields.length > 0) {
        fields.push({
            name: 'fields',
            admin: {
                style: {
                    borderBottom: 0,
                    borderTop: 0,
                    margin: 0,
                    padding: 0
                }
            },
            fields: Array.isArray(customFieldSchema) ? customFieldSchema.concat(extraFields) : extraFields,
            type: 'group'
        });
    }
    return fields;
}
